packages/frontend/src/contexts/AppContext/useTxHistory.ts

```diff
@@ -25,86 +27,200 @@ const cacheKey = 'recentTransactions:v000'
+    setTransactions(prevTransactions => {
+      if (!prevTransactions) return []
+      const filtered = filterByHash(prevTransactions, tx.hash)
+      return sortByRecentTimestamp(filtered).slice(0, 4)
+    })
+  }
+
+  function updateTransaction(tx: Transaction, updateOpts: UpdateTransactionOptions, matchingHash?: string) {
+    setTransactions(prevTransactions => {
+      if (!prevTransactions) return []
+      
+      // deep clone to avoid mutating state directly
+      const customizer = (value) => {
+        if (_.isFunction(value)) {
+          return value
+        }
       }
-      filterSortAndSetTransactions(tx, transactions, matchingHash || tx.hash)
-    },
-    [transactions]
-  )
-
-  // this will make sure to update in local storage the updated pending status,
-  // so it doesn't show as pending indefinitely on reload.
-  // This wouldn't be an issue if useEffect worked properly with array nested
-  // of nested objects and it detected property changes.
-  useEffect(() => {
-    transactions?.forEach(tx => {
-      const oldPending = tx.pending
-      const oldPendingDC = tx.pendingDestinationConfirmation
-      const cbPending = (pending: boolean) => {
-        if (oldPending !== pending) {
-          updateTransaction(tx, { pending })
+
+      const clonedTxs = _.cloneDeepWith(prevTransactions, customizer)
+      const targetTx = find(clonedTxs, ['hash', matchingHash || tx.hash])
+
+      if (targetTx) {
+        for (const key in updateOpts) {
+          targetTx[key] = updateOpts[key]
         }
       }
-      const cbPendingDC = (pendingDestinationConfirmation: boolean) => {
-        if (oldPendingDC !== pendingDestinationConfirmation) {
-          updateTransaction(tx, { pendingDestinationConfirmation })
+      
+      return sortByRecentTimestamp(clonedTxs).slice(0, 4)
+    })
+  }
+
+  // stores hashes for either pending origin or destination confirmation to prevent redundant listeners
+  const listenerSet = useRef(new Set())
+
+  const listenForOriginConfirmation = async (tx) => {
+    try {
+      tx.provider.once(tx.hash, transaction => {
+        listenerSet.current.delete(tx.hash)
+        updateTransaction(tx, { pending: false })
+      })
+    } catch (e) {
+      console.error('Error with origin transaction listener:', e)
+      listenerSet.current.delete(tx.hash)
+    }
+  }
+
+  // stores setInterval IDs -- may be cleared to stop polling
+  const intervalRefs = useRef({})
+  // stores setTimeouts to limit polling to one hour
+  const timeoutRefs = useRef({})
+
+  const POLLING_INTERVAL = 15000
+  const POLLING_TIMEOUT = 3600000
+
+  const getBondedTxHash = (tx) => {
+    return new Promise((resolve, reject) => {
+      const network = (typeof process !== 'undefined' && process.env && process.env.REACT_APP_NETWORK) ? process.env.REACT_APP_NETWORK : 'goerli'
+      const explorerAPIUrl = `https://${network === 'goerli' && 'goerli-'}explorer-api.hop.exchange/v1/transfers?transferId=${tx.hash}`
```

Comment:
suggestion: use the sdk to get the transfer status to avoid duplicated logic

```suggestion
const response = await sdk.getTransferStatus(tx.hash)
```